text/0000-event-processor-sdk.md

```diff
@@ -0,0 +1,232 @@
+data off-chain, but the current SDK does not provide a straight-forward way to
+subscribe and receive events.  This API provides an analogous framework to the
+transaction processors by removing the low-level transport and messaging
+details.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+The Event Processor SDK provides APIs for generating subscriptions, a handler
+interface for events, and an event processor for managing messaging and
+dispatching events to handlers.
+
+Subscriptions are generated through a subscription builder.  The protocol
+requirements and format are isolated behind this abstraction.  The builder is
+used to specify the event types and their filters.
+
+The event handler interface handles a subset of events received.  It specifies
+the names of events of interest.  It is provided the events of interest by the
+event processor.  Each handler should manage its own state, such as connections
+to databases or log files.
+
+The event processor takes a validator endpoint, a set of subscriptions, and a
+set of event handlers.  The event processor manages connecting to the validator,
+submitting the subscriptions, and receiving the event messages.  As events are
+received, the event processor finds the appropriate handlers for the given set
+of events, and provides the events of interest to each handler.
+
+Unlike transaction handlers, there may be multiple event handlers that are
+interested in specific event types.  For example, there may be a subscription
+for `sawtooth/block-commit`, `sawtooth/state-delta`, and `my-app/custom-event`.
+A handler may be created that handles the `block-commit` and `state-delta`
+events, and writes the state values to a
+[slowly-changing-dimensions](https://en.wikipedia.org/wiki/Slowly_changing_dimension#Type_2:_add_new_row)
+database.  A second handler may handle the `block-commit` and `custom-event` and
+forward the values to an external queue system, such as Apache Kafka.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## Subscription Builders
+[subscription-builders]: #subscription-builders
+
+The subscription builder API provides a clean abstraction for specifying
+Sawtooth event subscriptions.  
+
+The following API is presented in Rust, but is applicable to all of the SDKs.
+The builders are not traits, in this case, but the implementation of the structs
+and the functions are omitted here.
+
+```rust
+impl SubscriptionBuilder {
+    fn with_last_known_block_ids(&mut self, block_ids: Vec<String>)
+        -> Self
+    { ... }
+
+    fn with_event(&mut self, event_filter: EventFilter)
+        -> Self 
+    { ... }
+
+    fn build(self)
+        -> Result<Subscription, SubscriptionBuildError>
+    { ... }
+}
+
+enum FilterConstraint {
+    Any,
+    All,
+}
+
+impl EventFilterBuilder {
+    fn with_event_type(&mut self, event_type: String)
+        -> Self 
+    { ... }
+    
+    fn with_regex_filter(
```

Comment:
[Thinking out loud]: Can this be trait object passed to a function like "with_filter(<filter-type>)", where <filter-type> is either regex or simple or something else coming up later.